### DISCUSSION TITLE
classes/class_optionbutton

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#63] [User-contributed notes]
# classes/class_optionbutton

Inherits: Button< BaseButton< Control< CanvasItem< Node< Object A button that brings up a dropdown with selectable options when pressed. Description: OptionButton is a type of button that brings up...

https://docs.godotengine.org/en/4.3/classes/class_optionbutton.html

**Comment by eeyzs1** [comment]
just have 1 question, how can I change the font size of item?

**Reply by Calinou** [reply]
There doesn't appear to be a way to change the font size of a *single* item within an OptionButton, but there is a `font_size` theme item that affects the OptionButton itself. To change the font size of all items displayed within the dropdown, you need to create a theme and override [PopupMenu's `font_size`](https://docs.godotengine.org/en/stable/classes/class_popupmenu.html#class-popupmenu-theme-font-size-font-size) theme item.

**Reply by eeyzs1** [reply]
cool, many thanks!

**Comment by MrButier** [comment]
Is there any way to modify the OptionButton's popup.maxsize other than having to create your own custom dropdown menu? I was running into issues setting the max height for the popup as none of the popup maxsize functions/variables would work. I was able to make my own class that extends from "Button" and built it by scratch but wondering if I created more work for no reason.

**Reply by Kolbias** [reply]
I used this code from another thread and it works. $YourOptionButton.get_popup().max_size = Vector2(9999, 200)

**Comment by LoshyDoshy** [comment]
Is there anyway to randomize the order the options appear?

**Reply by Calinou** [reply]
This works on my end (attach this script to the OptionButton node):

```gdscript
extends OptionButton

func _ready():
	# Copy current items to dictionary.
	var items_copy: Array[Dictionary] = []
	for item_idx in item_count:
		items_copy.push_back({
			id = get_item_id(item_idx),
			icon = get_item_icon(item_idx),
			metadata = get_item_metadata(item_idx),
			text = get_item_text(item_idx),
			tooltip = get_item_tooltip(item_idx),
			disabled = is_item_disabled(item_idx),
			separator = is_item_separator(item_idx),
		})

	items_copy.shuffle()

	# Reset items and add new shuffled items.
	clear()
	for item in items_copy:
		if item.separator:
			add_separator(item.text)
		else:
			add_icon_item(item.icon, item.text, item.id)
		set_item_metadata(-1, item.metadata)
		set_item_tooltip(-1, item.tooltip)
		set_item_disabled(-1, item.disabled)


func _on_item_selected(index: int) -> void:
	# When querying the item's index after selection,
	# get its original ID so you know which item was actually selected
	# regardless of order.
	print(get_item_id(index))
```

Run 1 | Run 2
-|-
![image](https://github.com/user-attachments/assets/d2c78f80-27fe-4eba-9bb8-84b10d3b5bef) | ![image](https://github.com/user-attachments/assets/c5c12d38-32e3-4cfd-ac62-95fb19f60cdc)

I've opened a pull request to add this code sample:

- https://github.com/godotengine/godot/pull/104131

**Reply by LoshyDoshy** [reply]
You're amazing, thanks! And mine is only showing the number value from my enum instead of the text, any idea why?

**Reply by LoshyDoshy** [reply]
Unless that's what the last print line is for

**Reply by Calinou** [reply]
Yes, it shows the numeric value of the item when you select it. You can query the item text using `get_item_text()` instead.

**Comment by ztoph** [comment]
On top of the documentation it says: "Note: The ID values used for items are limited to 32 bits, not full 64 bits of int. This has a range of -2^32 to 2^32 - 1, i.e. -2147483648 to 2147483647"
My calculator says that 2^31 - 1 is 2147483647 and 2^32 - 1 is 4294967295 !

**Comment by Oddness-1** [comment]
So, got a bit of a problem when using _ValidateProperty to hide the options for the OptionButton. If I set "item_count" to PropertyUsageFlags.EditorOnly, that breaks the item list that's usually there, and I just see a list which seems to come from the underlying 'Popup'.

Is there a good way to prevent manual changes to the OptionButton's item list in the inspector? I'm modifying them programmatically (filling from an enum), and just using the other functionality which comes with the OptionButton.

To Replicate:
``` cs
public override void _ValidateProperty(Dictionary property)
    {
        var name = property["name"].AsStringName();
        if (name == "item_count" || name.ToString().StartsWith("item_"))
        {
            property["usage"] = (int)PropertyUsageFlags.NoEditor;
        }
    
        base._ValidateProperty(property);
    }
```

**Comment by RainbowAlpacas** [comment]
Is there any way to access the mouse_enter and mouse_exit of individual item in the popup? They react to hovering like buttons do but there doesn't seem like there is a way to get the signal or item themselves. The item_focus only works for traversing the options using ui_up or ui_down but I want to use the mouse hover.

**Comment by hfblazquez003** [comment]
Is there any way to modify via Themes or any other thing the options which are displayed?? For what I have seen, you can modify the main button which displays the options, but not the options themselves...

**Reply by calham21** [reply]
was having trouble with this just now! If you are using a theme, the Option Button menu is classified as a PopupMenu. Just add that to the theme you are using and you should be able to customize it!

**Reply by hfblazquez003** [reply]
Wow, Didnt even know you could add that... Thank you so much!!!!

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot/pull/118715

**Comment by Danster21** [comment]
Is there any way to change the order of text and icon? Currently if you set both to be in the Center, the text will display over the icon. I'd like to have a way to set the text to be behind the icon, or better yet, a way to order both of them independently so I could add additional nodes between them.

**Reply by Calinou** [reply]
As far as I know, this isn't supported yet. Please open a [proposal](https://github.com/godotengine/godot-proposals) describing the use case if you'd like this to be implemented :slightly_smiling_face:

**Comment by lupuchard** [comment]
This documentation says the ID has a range of -2147483648 to 2147483647, but this is misleading because it doesn't seem to be possible to have an ID of -1 (add_item uses -1 as the default and will set the ID to the index).
